### ISSUE TITLE
Allow internal dependents to not be automatically bumped

### ISSUE THREAD

**Issue by mondash** [atlassian/changesets#568]
## Affected Packages

First time issue submitter here, so I'm not 100% sure, but I think only these packages would be affected:
- `apply-release-plan`
- `assemble-release-plan`
- `cli`
- `config`
- `types`

## Problem

I'll construct a contrived example here to illustrate my real-world problem.

I've got a monorepo set up with several shared libraries, `foo@1.0.0` and `bar@1.0.0`, to be used in other applications. These libraries may have internal dependencies. So let's say that `foo@1.0.0` has a dependency on `bar@^1.0.0`. I make a breaking change to `bar@1.0.0` and create a major version changeset bumping to `bar@2.0.0`.

The next time I run `yarn changeset version` (or the changeset action runs in CI), in addition to bumping `bar@2.0.0`, `foo@1.0.0` is automatically bumped to `foo@1.0.1`, and its internal dependency is bumped to `bar@^2.0.0`. As far as I understand, this is intended behavior.

But what if I needed to make some changes to the `foo` library before it can use `bar@2.0.0`? Sure, I could make those changes before releasing `bar@2.0.0` and publish both releases together, but should the release of `bar@2.0.0` be blocked by the dependent `foo` library? This problem compounds if there are many packages which all depend on `bar@1.0.0`, such that coordinating the update of each dependent to be compliant with the new version of `foo@2.0.0` _before even releasing_ `foo@2.0.0` becomes near impossible.

Here's my `.changesets/config.json` for reference:

```json
{
  "$schema": "https://unpkg.com/@changesets/config@1.6.0/schema.json",
  "changelog": "@changesets/cli/changelog",
  "commit": false,
  "linked": [],
  "access": "restricted",
  "baseBranch": "main",
  "updateInternalDependencies": "patch",
  "ignore": []
}
```

## Proposed solution

What I am suggesting is a new config option (perhaps `"updateInternalDependencies": "never"` or something) which would no longer update internal dependencies to a package when a new version of it is published. This means that each package would need to manually update it's internal dependencies, but it would allow the releases of packages with dependents to be decoupled from the releases of their dependents.

I've sorted this out in theory in my TS monorepo. Since yarn workspaces will install the outdated versions of the internal dependencies in the packages' `node_modules`, I used path aliases to point to either the internal dependencies location in that package's `node_modules` or to that packages `src` folder or to . This handles the case where we're using the current version of the internal dependency _and_ the case where we're using an outdated version.

 As long as this is not a default option, this would not be a breaking change, and it would greatly improve the workflow of my monorepo setup. If I were more familiar with the source code and had the bandwidth, I'd take a crack at this myself. If this is a feature request you all would consider, I'd greatly appreciate it! Thanks a lot for your work on this project!! ❤️

**Comment by Andarist** [comment]
> should the release of bar@2.0.0 be blocked by the dependent foo library? 

From my experience - yes. This is inconvenient but otherwise, you end up with a bunch of different problems so this is the easiest, most maintainable, approach in the long run. Consider also this - how do you run tests correctly? In monorepo setup it's the most common scenario to test against the current versions of all packages. So in a sense, you already need to upgrade your `foo` to be compatible with the new `bar` as otherwise, your tests would fail.

From the latter description, I understand that you actually install outdated versions somehow, and thus this mentioned problem doesn't apply to you but (from my experience) this is a less popular choice.

I believe though that the use case that you mention is already covered by us but not yet sure how satisfying it will be for your setup. There is `bumpVersionsWithWorkspaceProtocolOnly` option that has been introduced in [this PR](https://github.com/atlassian/changesets/pull/507). It would require you to use the `workspace:` protocol for your monorepo dependencies but it would also allow you to cleanly opt-out from that - allowing you to use "outdated" versions on demand.

**Comment by mondash** [comment]
@Andarist This setting did the job for me. Thanks a lot for the speedy response!

I suppose that I'm not using a traditional monorepo setup (I'd call it a multipackage repo), so this will let me co-locate all my libraries while still having the versioning function as if the packages were hosted in their own repos.

I didn't think that the config option you mentioned was available to me as it doesn't appear to be included in the `./changeset/config.json` [schema](https://github.com/atlassian/changesets/blob/main/packages/config/schema.json). I'll leave this issue open for you to close for that reason.

**Comment by Andarist** [comment]
It seems that this option was not documented at all - as it's fairly new. I've prepared a PR which adds minima docs for it here: https://github.com/atlassian/changesets/pull/507 

Does this satisfy your use case completely? Or does something stay unresolved?

**Comment by mondash** [comment]
@Andarist Functionally this does satisfy my use case, yes, although I have found one small issue which looks like it was left over from #507.

When running `yarn changeset`, I'm getting a warning about internal dependencies to outdated versions.

>  Package "foo" must depend on the current version of "bar": "2.0.0" vs "^1.0.0"

It doesn't prevent me from creating new changesets, but I think this warning should be suppressed when `bumpVersionsWithWorkspaceProtocolOnly` is `true` and the dependency is not using the workspace protocol since using an outdated dependency here is allowed. Let me know what you think.

Thanks again for your help on this! 😄

**Comment by mondash** [comment]
@Andarist I have just found another unresolved issue (semi-related to the above mentioned issue). When using the changesets action (@master), I am still getting a changelog entry for bumping the internal dependency of `foo` on `bar`, even though it didn't actually happen. 

Ex:
```markdown
<!-- packages/foo/CHANGELOG.md -->

## 2.0.0

### Major Changes

- xxxxxxx: Some more setup

### Patch Changes

- Updated dependencies [xxxxxxx]
  - bar@2.0.0

```

**Comment by Andarist** [comment]
> It doesn't prevent me from creating new changesets, but I think this warning should be suppressed when bumpVersionsWithWorkspaceProtocolOnly is true and the dependency is not using the workspace protocol since using an outdated dependency here is allowed. Let me know what you think.

Yes, definitely this should be allowed.

> Perhaps the wording could be updated to saw "All workspace dependents..." or something like that.

Feel free to prepare a PR for this change.

> When using the changesets action (@master), I am still getting a changelog entry for bumping the internal dependency of foo on bar, even though it didn't actually happen.

Would be great if you could prepare a repro case or write a failing test in our repo - this would allow me to get to fixing this much quicker.

**Comment by antoine-coulon** [comment]
Hello @Andarist, 

I'm not sure if that issue is still a thing, however I'm encountering the same problem as mentioned just above there https://github.com/antoine-coulon/skott/pull/109

Basically I'm using a pnpm workspace where I'm actually not specifying any dependents bump, but it turns out that it's being added in the changelog anyway, even though the package itself is not being patch bumped (`skott-webapp` is already 1.0.6):

```
Patch Changes
https://github.com/antoine-coulon/skott/pull/108 [5f1d31b](https://github.com/antoine-coulon/skott/commit/5f1d31ba427d338bb9fee599ee2902e9f4e7cd96) Thanks [@antoine-coulon](https://github.com/antoine-coulon)! - Allow web application to be automatically opened on WSL via wslview. Better error handling when the application can't be automatically opened.

Updated dependencies []:

skott-webapp@1.0.6
```

Is it the expected behavior? It seems that `updateInternalDependencies` is true by default and can't be disabled, so maybe it's related? If the changelog write is not supposed to occur (for `skott-webapp`), I can try to provide a failing test + fix there. 

Thanks!

**Comment by jsudelko** [comment]
I'm trying to figuring out how the Changesets workflow should look like in a monorepo where dependencies are vendored through NPM and not `workspace:^`.

I'm in the situation where I needed to make some changes to the `foo` library before it can use `bar@2.0.0` and I need make those changes before releasing `bar@2.0.0` so that both `foo` and `bar` releases are published together. How does that work if the packages are or NPM and not using `workspace:^` versions since `bar` isn't published yet?

The only think I can think of is to locally change `foo` to a `workspace:^` instead of `bar@1.0.0`  and make the changes, then change foo back to `bar@1.0.0` so that Changesets can bump it accordingly. But in the meantime `foo` is using unpublished changes in "bar" so tests, type-checks etc. fail until after publish where they'll eventually succeed..

What am I missing? The fact that issue doesn't seem have much traffic seems makes me think I'm missing something straight-forward.
